### PR TITLE
[Feature] Replace column-parallel LM head with row-parallel LM head for DeepSeek v3

### PR DESCRIPTION
**Pull request by nbabinTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/43131

### Problem description
Current LM head uses column-parallel: all-gather input across TP to get full emb_dim, then matmul with vocab-sharded weight `[emb, vocab/tp]`. Output is TP-sharded on vocab, requiring concat to reconstruct full vocab for sampling. Alternative approach is to keep input TP-sharded `[TILE, emb/tp]`, use emb-sharded weight `[emb/tp, vocab]`, matmul gives partial sum `[TILE, vocab]`, then all-reduce across TP. Output is replicated with full vocab — no concat needed for sampling.

### What's changed
Implemented row-parallel LM head alongside the existing column-parallel, benchmarked both on Blackhole 8×4 at DSv3 dimensions (emb=7168, vocab=129280), and identified that the dominant cost were not the CCL ops but the device-to-host transfer of logits.

Column-parallel uses the following sequence of ops:
```
input -> All Gather -> Matmul -> output -> device-to-host 
```

Row-parallel uses:
```
input -> Matmul -> All Reduce (Reduce Scatter + All Gather) -> output -> device-to-host
```

We measured the cost of CCL ops and device to host transfer in column-parallel and row-parallel cases and got the following numbers:

| | Column-parallel | Row-parallel |
| :------- | :-------: | :------: |
| CCL ops | 38 us | 184 us + 165 us = 349 us |
| Optimized device to host | 13.60 ms | 6.67 ms |
| Total | 13.64 ms | 7.02 ms |

We also managed to optimize device to host transfers by reducing the number of chips we read from. In case of column-parallel approach, we now read only the row of devices that have the token of interest - tensors from those 4 devices are transferred to host where they are concatenated. In row-parallel case, we need to read from only one device in row that contains the token of interest. We read from only one device as data is replicated on all devices across columns.

Full analysis is available on the ticket: https://github.com/tenstorrent/tt-metal/issues/43131

Code changes:
* `models/demos/deepseek_v3_d_p/tt/tt_lm_head.py` — added a mode parameter (`is_column_parallel` parameter, default to `False` as the results show that it is more efficient)
* `models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py` — parametrized it over both modes
* `models/demos/deepseek_v3_d_p/tt/moe/tt_prefill_transformer.py` — propagates `lm_head_is_column_parallel` flag through both `check_cache_complete()` and `__init__` so the cache check and the LM head construction reference the same mode-specific cache files.

### DeepSeek Prefill CI

- [ ] [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:nbabin/lm_head_row_parallel)
- [ ] [(Blackhole) e2e tests](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-e2e-tests.yaml?query=branch:nbabin/lm_head_row_parallel)
- [ ] [(Galaxy) DeepSeek Prefill tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-deepseek-prefill-tests.yaml?query=branch:nbabin/lm_head_row_parallel)
- [ ] [(T3K) T3000 e2e tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-e2e-tests.yaml?query=branch:nbabin/lm_head_row_parallel)
- [ ] [(Release) Model Status for Console Deployment](https://github.com/tenstorrent/tt-metal/actions/workflows/demo-sp-release.yaml?query=branch:nbabin/lm_head_row_parallel)

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:nbabin/lm_head_row_parallel)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:nbabin/lm_head_row_parallel)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:nbabin/lm_head_row_parallel)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:nbabin/lm_head_row_parallel)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:nbabin/lm_head_row_parallel)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:nbabin/lm_head_row_parallel)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:nbabin/lm_head_row_parallel)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:nbabin/lm_head_row_parallel)

### CODE REVIEW

File: models/demos/deepseek_v3_d_p/tt/tt_lm_head.py [hunk 1]
```diff
@@ -164,20 +194,25 @@ def __init__(
...
+                                TP-sharded on vocab). False (default) - row-parallel
+                                (matmul partials + all_reduce, output TP-replicated).
         """
         super().__init__()
         self.mesh_device = mesh_device
         self.emb_dim = emb_dim
         self.vocab_size = vocab_size
         self.num_devices = mesh_device.get_num_devices()
         self.sp_factor = mesh_device.shape[0]
+        self.dp_factor = mesh_device.shape[1]
```
Comment (ipotkonjak-tt) [inline models/demos/deepseek_v3_d_p/tt/tt_lm_head.py]: this should be tp factor? but seems unused

File: models/demos/deepseek_v3_d_p/tt/tt_lm_head.py [hunk 2]
```diff
@@ -315,48 +359,79 @@ def forward(self, x: ttnn.Tensor, global_token_id: int) -> tuple[ttnn.Tensor, tu
...
+            # Column-parallel: All-gather x to get full emb_dim (replicated across TP axis)
+            # ========================================
+            # Input x is sharded: (dispatch_group_size/axis0, seq_len_per_chip, emb_dim/axis1)
+            # Both shared_expert and dispatch need full emb_dim, so all-gather first
+            # Only needed if there are multiple devices in TP axis (axis 1)
+            if tp_size > 1:
+                x_full = ttnn.all_gather(
+                    x,
+                    dim=-1,  # Gather along emb_dim
+                    cluster_axis=1,  # Gather across axis 1 (TP axis)
```
Comment (ipotkonjak-tt) [inline models/demos/deepseek_v3_d_p/tt/tt_lm_head.py]: avoid hardcoding, keep the info about which axis is TP

File: models/demos/deepseek_v3_d_p/tt/tt_lm_head.py [hunk 3]
```diff
@@ -315,48 +359,79 @@ def forward(self, x: ttnn.Tensor, global_token_id: int) -> tuple[ttnn.Tensor, tu
...
+        """
+        Read only from the row of devices that holds the target token's logits.
+
+        - row mode: TP slices are replicated, so read only from one device (the full vocab
+                    sits on every device in the row).
+        - column mode: TP slices are vocab/tp shards, so read from all devices in the row and
+                       concatenate on host to assemble the full vocab.
+        """
+        ttnn.synchronize_device(self.mesh_device)
+        tp = self.mesh_device.shape[1]
```
Comment (ipotkonjak-tt) [inline models/demos/deepseek_v3_d_p/tt/tt_lm_head.py]: seems like you should have tp_facor and tp_axis saved in the module. They should be created on init

File: models/demos/deepseek_v3_d_p/tt/moe/tt_prefill_transformer.py [hunk 4]
```diff
@@ -62,6 +63,10 @@ def check_cache_complete(
             num_layers: Number of transformer layers
             experts_per_chip: Number of routed experts per chip (default: 8)
             first_k_dense: Number of initial dense (non-MoE) layers (default: 3)
+            lm_head_is_column_parallel: Selects the LM head cache filename to look for.
+                Must match the value passed to the TtPrefillTransformer constructor so
+                the cache check and the actual load reference the same files.
+                Defaults to False (row-parellel mode) to match TtLMHead's default.
```
[the remark this answers is not in the document]
Comment (fbajraktariTT) [reply models/demos/deepseek_v3_d_p/tt/moe/tt_prefill_transformer.py]: Can we have the same copy in memory for both row/column parallel and just use different mesh_mappers for each?

File: models/demos/deepseek_v3_d_p/tt/tt_prefill_transformer.py [hunk 5]
```diff
@@ -114,6 +119,7 @@ def __init__(
         shared_expert_activations_dtype=ttnn.bfloat16,
         shared_expert_weights_dtype=ttnn.bfloat8_b,
         weight_cache_path: Optional[Path] = None,
+        lm_head_is_column_parallel: bool = False,
```
Comment (fbajraktariTT) [inline models/demos/deepseek_v3_d_p/tt/tt_prefill_transformer.py:117]: Is the fastest case a default?

Comment (nbabinTT) [reply models/demos/deepseek_v3_d_p/tt/tt_prefill_transformer.py:117]: Yes, it is.

File: models/demos/deepseek_v3_d_p/tt/tt_lm_head.py (hunk 1 above)
Comment (nbabinTT) [reply models/demos/deepseek_v3_d_p/tt/tt_lm_head.py]: Fixed it

File: models/demos/deepseek_v3_d_p/tt/tt_lm_head.py (hunk 2 above)
Comment (nbabinTT) [reply models/demos/deepseek_v3_d_p/tt/tt_lm_head.py]: Fixed it

File: models/demos/deepseek_v3_d_p/tt/tt_lm_head.py (hunk 3 above)
Comment (nbabinTT) [reply models/demos/deepseek_v3_d_p/tt/tt_lm_head.py]: Fixed it

**Review by ipotkonjak-tt — APPROVED** [review]
